include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -0,0 +1,378 @@
+
+            for (uint64_t i = 1; i <= parameter.max_temperature * 100; i++)
+            {
+                temp_values.push_back(static_cast<double>(i) / 100.0);
+            }
+
+            // All cells of the given layout are collected.
+            std::vector<typename Lyt::cell> all_cells{};
+            all_cells.reserve(layout.num_cells());
+
+            layout.foreach_cell([&all_cells](const auto& c) { all_cells.push_back(c); });
+
+            // The goal is to sort the cells from left to right and top to bottom.
+            std::sort(all_cells.begin(), all_cells.end());
+
+            // The energy distribution of the physically valid charge configurations for the given layout is determined.
+            const auto distribution = energy_distribution(stats_exhaustive.valid_lyts);
+
+            std::vector<int64_t> output_bits_index{};
+            std::stringstream    output_bits{};
+
+            if (parameter.truth_table.num_vars() == 3)
+            {
+                if (parameter.truth_table.num_bits() == 8)
+                {
+                    output_bits_index = {-4, -3};  // double wire, cx, etc.
+                    output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit * 2 + 1));
+                    output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit * 2));
+                    std::cout << output_bits.str() << std::endl;
+                }
+            }
+
+            else if (parameter.truth_table.num_vars() == 1 && parameter.truth_table.num_bits() == 2)
+            {
+
+                output_bits_index = {-2};
+                output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit));
+            }
+
+            else if (parameter.truth_table.num_vars() == 2)
+            {
+                if (parameter.truth_table.num_bits() == 4 &&
+                    parameter.truth_table != create_fan_out_tt())  // and, or, nand, etc.
+                {
+                    output_bits_index = {-2};
+                    output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit));
+                }
+                else
+                {
+                    output_bits_index = {-4, -3};  // fo2.
+                    output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit * 2 + 1));
+                    output_bits << std::to_string(kitty::get_bit(parameter.truth_table, parameter.input_bit * 2));
+                }
+            }
+
+            // Output cell(s) is/are collected.
+            std::vector<typename Lyt::cell> output_cells;
+            output_cells.reserve(output_bits_index.size());
+
+            std::transform(output_bits_index.begin(), output_bits_index.end(), std::back_inserter(output_cells),
+                           [&all_cells](const auto& index) { return *(all_cells.end() + index); });
```

Comment:
I don't understand what is happening here. It seems dangerous to add an index to an `end` iterator and then access the resulting field. Why is this safe? Could you add a comment here, please?